server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java

```diff
@@ -129,21 +135,83 @@ void testGetClientAuthenticationParametersNullJwtClientConfiguration() {
     assertEquals(params, jwtClientAuthentication.getClientAuthenticationParameters(params, config));
   }
 
-  private void mockKeyInfoService() {
+  @Test
+  void testGetClientAssertionUnknownSingingKey() throws ParseException {
+    // Given
+    HashMap customClaims = new HashMap<>();
+    customClaims.put("kid", "wrong-key-id");
+    config.setJwtClientAuthentication(customClaims);
+    // When
+    assertThrowsExactly(BadCredentialsException.class, () -> jwtClientAuthentication.getClientAssertion(config));
+  }
+
+  @Test
+  void testGetClientAssertionCustomSingingKey() throws ParseException {
+    // Given
+    mockKeyInfoService("myKey", JwtHelperX5tTest.CERTIFICATE_1);
+    HashMap customClaims = new HashMap<>();
+    customClaims.put("kid", "myKey");
+    config.setJwtClientAuthentication(customClaims);
+    MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+    // When
+    params = jwtClientAuthentication.getClientAuthenticationParameters(params, config);
+    // Then
+    assertTrue(params.containsKey("client_assertion"));
+    assertTrue(params.containsKey("client_assertion_type"));
+    JWSHeader header = validateClientAssertionOidcComplaint((String) params.get("client_assertion").get(0));
+    assertEquals("myKey", header.getKeyID());
+    assertNull(header.getJWKURL());
+  }
+
+  @Test
+  void testGetClientAssertionCustomSingingKeyButNoCertificate() throws ParseException {
+    // Given
+    mockKeyInfoService("myKey", null);
+    HashMap customClaims = new HashMap<>();
+    customClaims.put("kid", "myKey");
+    config.setJwtClientAuthentication(customClaims);
+    MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+    // When
+    params = jwtClientAuthentication.getClientAuthenticationParameters(params, config);
+    // Then
+    assertTrue(params.containsKey("client_assertion"));
+    assertTrue(params.containsKey("client_assertion_type"));
+    validateClientAssertionOidcComplaint((String) params.get("client_assertion").get(0));
```

Comment:
It seems that line 179 is unnecessary as the same call is made in line 180.

Reply:
done